src/daft-core/src/array/ops/hash.rs

```diff
@@ -23,73 +27,121 @@ impl<T> DataArray<T>
 where
     T: DaftPrimitiveType,
 {
-    pub fn hash(&self, seed: Option<&UInt64Array>) -> DaftResult<UInt64Array> {
+    pub fn hash(
+        &self,
+        seed: Option<&UInt64Array>,
+        hash_function: HashFunctionKind,
+    ) -> DaftResult<UInt64Array> {
+        self.hash_with_specified_algorithm(seed, hash_function)
+    }
+    pub fn hash_with_specified_algorithm(
+        &self,
+        seed: Option<&UInt64Array>,
+        _hash_function: HashFunctionKind,
+    ) -> DaftResult<UInt64Array> {
         let as_arrowed = self.as_arrow();
-
         let seed = seed.map(|v| v.as_arrow());
-
-        let result = kernels::hashing::hash(as_arrowed, seed)?;
-
+        let result = kernels::hashing::hash(as_arrowed, seed, _hash_function)?;
         Ok(DataArray::from((self.name(), Box::new(result))))
     }
 }
 
 impl Utf8Array {
-    pub fn hash(&self, seed: Option<&UInt64Array>) -> DaftResult<UInt64Array> {
+    pub fn hash(
+        &self,
+        seed: Option<&UInt64Array>,
+        hash_function: HashFunctionKind,
+    ) -> DaftResult<UInt64Array> {
+        self.hash_with_specified_algorithm(seed, hash_function)
+    }
```

Comment:
For the sake of compatibility, let's either keep the hash signature the same, passing in xxhash as the algorithm. Or only having one method.

I suggest updating the code to,

```rust
pub fn hash(&self, seed: Option<&Uint64Array>) -> DaftResult<UInt64Array> {
  self.hash_with(seed, HashFunctionKind::XXHash)
}

pub fn hash_with(&self, seed: Option<&UInt64Array>, hash_function: HashFunctionKind) -> ....

```

Reply:
Thanks @rchowell, seems like we still need a way to pass in the user-specified hash_function to hash() function? This suggested change will only use XXHash as default hash algorithm but not sure if there's a way to specify other hash algorithm like SHA1 without changing the originally hash signature?

Reply:
We only need to pass the user-defined hash function in the implementation of the `Expression` function implementation. The existing `hash()` can remain the same.